### PR TITLE
Support setting "defined tags" in data science jobs.

### PR DESCRIPTION
**Pull request by qiuosier**
- Adding `with_defined_tag()` to job runtime.
- Adding `with_freeform_tag()` and `with_defined_tag()` job infrastructure.
- Tags defined in runtime will overwrite the ones defined in infrastructure.
- Tags defined in runtime will also be added to job run.
- Tags defined in infrastructure will be added to job only.

### CODE REVIEW

**Review by mrDzurb — APPROVED** [review]
Could you also update docstring examples for job and jobrun and add tags lines there?

File: tests/unitary/default_setup/catalog/test_catalog_project.py [hunk 1]
```diff
@@ -61,14 +61,14 @@ class ProjectCatalogTest(unittest.TestCase):
 
     with patch.object(auth, "default_signer"):
         with patch.object(oci_client, "OCIClientFactory"):
-            pc = ProjectCatalog()
-            pc.ds_client = MagicMock()
-            pc.identity_client = MagicMock()
-
             project_id = "ocid1.projectcatalog.oc1.iad.<unique_ocid>"
```
Comment (ahosler) [inline tests/unitary/default_setup/catalog/test_catalog_project.py:64]: Should we have something here like: `project_id = os.environ.get("NB_PROJECT_OCID",  "ocid1.projectcatalog.oc1.iad.<unique_ocid>")`

**Review by ahosler — APPROVED** [review]
Great to see this 👍

File: tests/unitary/default_setup/catalog/test_catalog_project.py (hunk 1 above)
Comment (qiuosier) [reply tests/unitary/default_setup/catalog/test_catalog_project.py:64]: It was like that in the unit tests. But I guess it does not matter. I am adding another PR to make unit tests have less dependency on environment variable.

**Review by mrDzurb — APPROVED** [review]

**Review by lu-ohai — APPROVED** [review]

**Review by lu-ohai — APPROVED** [review]

**Review by mrDzurb — APPROVED** [review]

**Pull request merged**